src/sparse/impl/KokkosSparse_twostage_gauss_seidel_impl.hpp

```diff
@@ -0,0 +1,919 @@
+        tic = timer.seconds ();
+        std::cout << std::endl << "TWO-STAGE GS::NUMERIC::INSERT LU TIME : " << tic << std::endl;
+        timer.reset();
+#endif
+
+        if (!(gsHandle->isTwoStage ())) {
+          // TODO: figure out how to use CuSPARSE
+          #if 0 // defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA)
+          using namespace KokkosSparse::Experimental;
+          sptrsv_symbolic (handle->get_gs_sptrsvL_handle(), rowmap_viewL, crsmatL.graph.entries, values_viewL);
+          sptrsv_symbolic (handle->get_gs_sptrsvU_handle(), rowmap_viewL, crsmatU.graph.entries, values_viewU);
+          #endif
+        }
+      }
+
+
+      /**
+       * Apply solve
+       */
+      template <typename x_value_array_type, typename y_value_array_type>
+      void apply (x_value_array_type localX, // in/out
+                  y_value_array_type localB, // in
+                  bool init_zero_x_vector = false,
+                  int numIter = 1,
+                  scalar_t omega = Kokkos::Details::ArithTraits<scalar_t>::one(),
+                  bool apply_forward = true,
+                  bool apply_backward = true,
+                  bool update_y_vector = true)
+      {
+        const_scalar_t one (1.0);
+        const_scalar_t zero (0.0);
+#ifdef KOKKOSSPARSE_IMPL_TIME_TWOSTAGE_GS
+        double tic;
+        Kokkos::Impl::Timer timer;
+        Kokkos::fence();
+        tic = timer.seconds ();
+#endif
+
+        //
+        auto *gsHandle = get_gs_handle();
+        bool two_stage = gsHandle->isTwoStage ();
+        GSDirection direction = gsHandle->getSweepDirection ();
+        if (apply_forward && apply_backward) {
+          direction = GS_SYMMETRIC;
+        } else if (apply_forward) {
+          direction = GS_FORWARD;
+        } else if (apply_backward) {
+          direction = GS_BACKWARD;
+        } else {
+          return;
+        }
+
+        // load auxiliary matrices from handle
+        auto localD = gsHandle->getD ();
+        auto crsmatL = gsHandle->getL ();
+        auto crsmatU = gsHandle->getU ();
+
+        // wratp A into crsmat
+        input_graph_t graphA (column_view, rowmap_view);
+        input_crsmat_t crsmatA ("A", num_rows, values_view, graphA);
+#ifdef KOKKOSSPARSE_IMPL_TIME_TWOSTAGE_GS
+        Kokkos::fence();
+        tic = timer.seconds ();
+        std::cout << std::endl << "TWO-STAGE GS::APPLY::CREATE CRS_A TIME : " << tic << std::endl;
+        timer.reset();
+#endif
+
+        // load auxiliary vectors
+        int nrows = num_rows;
+        int nrhs = localX.extent (1);
+        gsHandle->initVectors (nrows, nrhs);
+        auto localR = gsHandle->getVectorR ();
+        auto localT = gsHandle->getVectorT ();
+        auto localZ = gsHandle->getVectorZ ();
+
+        // outer Gauss-Seidel iteration
+        int NumSweeps = numIter;
+        int NumInnerSweeps = gsHandle->getNumInnerSweeps ();
+        if (direction == GS_SYMMETRIC) {
+          NumSweeps *= 2;
```

Comment:
Is this consistent with the other GS algorithms? In that case we might want to communicate that, I know a few MueLu developers that might not be aware of that implementation choice ; )

Reply:
@lucbv It's consistent because here, each loop iteration does a forward or backward sweep, but not both. So doubling the loop count for symmetric is the right thing to do.

Reply:
OK good with me